### PR TITLE
op-node: Fixed Integers in channel frame header

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
This switches the channel frame header over to using fixed integers. This avoids a circular
dependency when filling the serialized frame up to a maximum size. The circular dependency
is caused by the frame data length being a `uvarint` which can vary in size when serialized.

**Metadata**
- Fixes ENG-2371
- Fixes ENG-2537

### CODE REVIEW

File: op-node/rollup/derive/channel_frame.go [hunk 1]
```diff
@@ -91,21 +77,20 @@ func (f *Frame) UnmarshalBinary(r ByteReader) error {
 	if err != nil {
 		return fmt.Errorf("error reading ID: %w", err)
 	}
-	f.ID.Time, err = binary.ReadUvarint(r)
-	if err != nil {
-		return fmt.Errorf("error reading ID.Time: %w", err)
+	if err := binary.Read(r, binary.BigEndian, &f.ID.Time); err != nil {
```
Comment (mslipper) [inline op-node/rollup/derive/channel_frame.go:80]: This is cool, didn't know about `binary.Read`. I've always used `binary.BigEndian.ReadUint64`, etc.

File: specs/derivation.md [hunk 2]
```diff
@@ -265,15 +265,23 @@ frame = channel_id ++ frame_number ++ frame_data_length ++ frame_data ++ is_last
...
+frame_number      = uint16
+frame_data_length = uint32
 frame_data        = bytes
 is_last           = bool
 ```
 
 > **TODO** replace `uvarint` by fixed size integers
 
+timestamp: 64 bits
+frame_number: 16 bits?
```
Comment (mslipper) [inline specs/derivation.md:278]: [nit] you can delete the **TODO** above.

**Review by mslipper — APPROVED** [review]

File: op-node/rollup/derive/channel_frame.go (hunk 1 above)
Comment (trianglesphere) [reply op-node/rollup/derive/channel_frame.go:80]: I can't actually find a `binary.BigEndian.ReadUint64` which works on readers, a lot of the binary encoding API works on buffers

File: specs/derivation.md [hunk 3]
```diff
@@ -270,22 +270,16 @@ frame_number      = uint16
...
-timestamp: 64 bits
-frame_number: 16 bits?
-frame_data_length: 32 bits?
-
-Fixed overhead: 32 + 8 + 2 + 4 + 1 = 47 bytes. Can get down to 16 + 6 = 22 bytes (128 bit UID, 16 bit #, 32 bit len)
+Where `uint64`, `uint32` and `uint16` are all big-endian unsigned integers.
+```
 
-Why fixed length: don't have circular dependency b/c have target data length
+All data in a frame is fixed-size, except the `frame_data`. The fixed overhead is `32 + 8 + 2 + 4 + 1 = 47 bytes`.
```
Comment (trianglesphere) [inline specs/derivation.md:277]: Out of scope for this PR, but I think we can cut this down to ~22 bytes

File: op-node/rollup/derive/channel_frame.go [hunk 4]
```diff
@@ -12,70 +12,56 @@ import (
 // but we leave space to grow larger anyway (gas limit allows for more data).
 const MaxFrameLen = 1_000_000
```
Comment (trianglesphere) [inline op-node/rollup/derive/channel_frame.go:13]:
```suggestion
const MaxFrameLen = 10_000_000
```

File: specs/derivation.md [hunk 5]
```diff
@@ -290,7 +292,7 @@ where:
       margin. (A soft constraint is not a consensus rule — nodes will accept such blocks in the canonical chain but will
       not attempt to build directly on them.)
 - `frame_number` identifies the index of the frame within the channel
-- `frame_data_length` is the length of `frame_data` in bytes
+- `frame_data_length` is the length of `frame_data` in bytes. It is capped to 1,000,000 bytes.
```
Comment (tynes) [inline specs/derivation.md:295]: I feel like this should be a power of two? Why this number?

Comment (trianglesphere) [reply specs/derivation.md:295]: This was proto's suggestion. I'd probably go a tad bigger, but I don't have a strong opinion on what it should be. We probably need to look into cannon's limits here

Comment (protolambda) [reply specs/derivation.md:295]: Any limit is fine for now, we're not hitting it for now. Doesn't have to be a power of 2. Feel free to change it to a power of 2 though (eth2 specs do it everywhere, except the 1MB network gossip limit, haha).

File: op-node/rollup/derive/channel_bank_test.go [hunk 6]
```diff
@@ -106,32 +108,28 @@ func (tf testFrame) Content() []byte {
...
 	require.NoError(bt.t, bt.cb.IngestData(data))
 }
 func (bt *bankTestSetup) ingestFrames(frames ...testFrame) {
-	data := []byte{DerivationVersion0}
+	data := new(bytes.Buffer)
+	data.WriteByte(DerivationVersion0)
 	for _, fr := range frames {
-		data = append(data, fr.Encode()...)
+		f := fr.ToFrame()
+		if err := (&f).MarshalBinary(data); err != nil {
```
Comment (protolambda) [inline op-node/rollup/derive/channel_bank_test.go:128]:
```suggestion
		if err := f.MarshalBinary(data); err != nil {
```
We can call methods with pointer receivers on symbols without doing this thing without negative effects I think.

**Review by protolambda — COMMENTED** [review]
looks good, but not approving just yet to avoid auto-merge, one nit style fix first.

**Review by protolambda — APPROVED** [review]

**Pull request merged**